ESQL: Refactor Greatest and Least functions to use evaluator map (#114036)

shinsj4653 opened the pull request:
# ESQL: Refactor Greatest and Least functions to use switch statements

## Description
This PR refactors the `Greatest` and `Least` functions to use Java switch expressions instead of Map-based evaluator lookup for better performance and code clarity. The changes include:

1. Replaced `EVALUATOR_MAP` with switch expressions in `toEvaluator` method
2. Added `isSupportedDataType` helper method using switch expression
3. Improved error messages with proper type validation
4. Added comprehensive type validation tests in `VerifierTests`

## Changes
- `Greatest.java`:
  - Replaced Map-based evaluator lookup with switch expression
  - Added `isSupportedDataType` method for type validation
  - Updated `toEvaluator` to use switch-based factory creation
  - Improved error handling and type checking

- `Least.java`:
  - Replaced Map-based evaluator lookup with switch expression  
  - Added `isSupportedDataType` method for type validation
  - Updated `toEvaluator` to use switch-based factory creation
  - Improved error handling and type checking

- `VerifierTests.java`:
  - Added `testGreatestLeastWithUnsupportedDataTypes()` test method
  - Tests unsupported data types (geo_point, cartesian_point)
  - Tests mixed type scenarios with proper error message validation
  - Ensures type consistency between `resolveType` and `toEvaluator`

## Testing

- `GreatestTests.java`, `LeastTests.java`
```bash
./gradlew :x-pack:plugin:esql:test --tests "*GreatestTests*" --tests "*LeastTests*" -x :libs:simdvec:compileMain21Java -x :libs:simdvec:compileMain22Java --console=plain
```
<img width="695" height="109" alt="image" src="https://github.com/user-attachments/assets/b0a894ef-a332-4336-aa5f-4197b9748104" />  

- `VerifierTests.testGreatestLeastWithUnsupportedDataTypes`
```bash
./gradlew :x-pack:plugin:esql:test --tests "VerifierTests.testGreatestLeastWithUnsupportedDataTypes" -x :libs:simdvec:compileMain21Java -x :libs:simdvec:compileMain22Java --console=plain --info
```
<img width="737" height="372" alt="image" src="https://github.com/user-attachments/assets/85fededd-54fa-43ff-ac7e-057aa7875971" />

### ✅ Greatest/Least Function Tests
- **GreatestTests**: All tests PASSED
- **LeastTests**: All tests PASSED
- Covers all supported data types and edge cases
- No regressions introduced

### ✅ VerifierTests - Type Validation
- **testGreatestLeastWithUnsupportedDataTypes**: PASSED
- Tests unsupported data types (geo_point, cartesian_point)
- Validates proper error messages for type mismatches
- Ensures consistency between `resolveType` and `toEvaluator`

### 🔍 Error Scenario Validation
The new type validation correctly rejects unsupported data types:
- `geo_point` → "Cannot use [geo_point] with function [Greatest]"
- `cartesian_point` → "Cannot use [cartesian_point] with function [Least]"
- Mixed types → Proper type mismatch error messages

**Note**: The error messages are validated by the test framework and don't appear in logs when tests pass, which is the expected behavior.

## Related Issues
Closes #114036

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Backport
- [ ] Backport to 8.19
- [ ] Backport to 8.18
- [ ] Backport to 8.17

shinsj4653 committed:
Refactor Greatest function to use evaluator map

This commit refactors the Greatest function to use an evaluator map for better
code organization. The changes include:
- Added EVALUATOR_MAP for evaluator factory mapping
- Updated toEvaluator to use map-based lookup
- Added NULL type validation
- Updated error messages to use getWriteableName()

The refactoring improves code readability and maintainability by:
- Replacing if-else chains with a map-based lookup
- Centralizing evaluator factory creation
- Adding consistent NULL type validation
- Using getWriteableName() for error messages

Relates to #114036

shinsj4653 committed:
Refactor Least function to use evaluator map

This commit refactors the Least function to use an evaluator map for better
code organization. The changes include:
- Added EVALUATOR_MAP for evaluator factory mapping
- Updated toEvaluator to use map-based lookup
- Added NULL type validation
- Updated error messages to use getWriteableName()

The refactoring improves code readability and maintainability by:
- Replacing if-else chains with a map-based lookup
- Centralizing evaluator factory creation
- Adding consistent NULL type validation
- Using getWriteableName() for error messages

Relates to #114036

shinsj4653 committed:
Add tests for Greatest and Least functions

This commit adds test cases for both Greatest and Least functions to verify:
- All supported data types (boolean, double, integer, long, keyword, text, version, ip, datetime, date_nanos)
- NULL value handling
- String comparisons
- Numeric comparisons
- Boolean operations
- Date/time comparisons

The tests ensure that:
- Both functions handle NULL values correctly
- Type validation works as expected
- Evaluator selection is correct for each data type
- Results are computed correctly for all supported types

Relates to #114036

shinsj4653 committed:
Fix PR number in commit messages (#128429)

shinsj4653 committed:
Update docs/changelog/128429.yaml

alex-spies reviewed (dismissed):
Hi @shinsj4653 and thank you very much for your contribution!

I'm sorry it took us so long to get back to you.

Unfortunately, this PR does not contain tests that demonstrate that the earlier type check actually happens and returns the expected error message. We often add cases to the `VerifierTests` in such situations.

shinsj4653 commented:
Hi @alex-spies, I appreciate your feedback. I'll add a test case for this issue using VerifierTests

shinsj4653 commented:
Hi @alex-spies @idegtiarenko,

Thank you for the valuable feedback! I've implemented the suggested changes:

✅ **idegtiarenko's feedback**: Refactored to use switch expressions instead of Map-based evaluator lookup
✅ **alex-spies's feedback**: Added comprehensive type validation tests in VerifierTests

The changes ensure:
- Better performance with switch expressions
- Proper type validation for unsupported data types
- Consistency between resolveType and toEvaluator

All tests are passing. PTAL and I'd appreciate your review! 🙏

ncordon commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -2590,6 +2590,118 @@ private void checkFullTextFunctionNullArgs(String functionInvocation, String arg
         );
     }
 
+public void testGreatestLeastWithUnsupportedDataTypes() {
```
This doesn't look correctly indented, which probably the CI will complain about. You should be able to format these files with:

```
./gradlew :x-pack:plugin:esql:spotlessApply 
```

ncordon commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -2590,6 +2590,118 @@ private void checkFullTextFunctionNullArgs(String functionInvocation, String arg
+        );
+    }
+
+    public void testMultiMatchFunctionNotAllowedAfterCommands() throws Exception {
```
Why are we adding tests for multimatch here?

ncordon reviewed (changes requested):
Thanks a lot for the contribution @shinsj4653!

> ✅ idegtiarenko's feedback: Refactored to use switch expressions instead of Map-based evaluator lookup

We have now duplicated the logic for knowing whether we support a type in `resolveType()` and `toEvaluator()` and we should not do that in my opinion. Otherwise, if somebody added a new type, they would have to add it to both places. Going with that approach misses the point of the issue https://github.com/elastic/elasticsearch/issues/114036, so I'm going to ask you to please revert to the evaluator map you used in https://github.com/elastic/elasticsearch/pull/128429/changes/95f84d28b019d07c1c648bf61f3c09fec75a4081

ncordon commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Least.java:
```diff
@@ -42,6 +47,17 @@ public class Least extends EsqlScalarFunction implements OptionalArgument {
+        DataType.DATE_NANOS, factories -> new LeastLongEvaluator.Factory(Source.EMPTY, factories),
+        DataType.IP, factories -> new LeastBytesRefEvaluator.Factory(Source.EMPTY, factories),
+        DataType.VERSION, factories -> new LeastBytesRefEvaluator.Factory(Source.EMPTY, factories)
+    );
```
This is not correct. We cannot pass `Source.EMPTY`.

For reference in https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/GreaterThan.java we have this code:

```java
    private static final Map<DataType, EsqlArithmeticOperation.BinaryEvaluator> evaluatorMap = Map.ofEntries(
        Map.entry(DataType.INTEGER, GreaterThanIntsEvaluator.Factory::new),
        Map.entry(DataType.DOUBLE, GreaterThanDoublesEvaluator.Factory::new),
        Map.entry(DataType.LONG, GreaterThanLongsEvaluator.Factory::new),
        Map.entry(DataType.UNSIGNED_LONG, GreaterThanLongsEvaluator.Factory::new),
        Map.entry(DataType.DATETIME, GreaterThanLongsEvaluator.Factory::new),
        Map.entry(DataType.DATE_NANOS, GreaterThanLongsEvaluator.Factory::new),
        Map.entry(DataType.KEYWORD, GreaterThanKeywordsEvaluator.Factory::new),
        Map.entry(DataType.TEXT, GreaterThanKeywordsEvaluator.Factory::new),
        Map.entry(DataType.VERSION, GreaterThanKeywordsEvaluator.Factory::new),
        Map.entry(DataType.IP, GreaterThanKeywordsEvaluator.Factory::new)
    );
```

and then we apply that to the source and the arguments:

```java
evaluatorMap.get(commonType).apply(source(), lhs, rhs);
```

ncordon commented:
buildkite test this please

shinsj4653 committed:
Address review feedback: Revert to Map-based evaluator and fix Source propagation

ncordon reviewed (commented):
Last suggested nit I am happy to approve and merge 😄

shinsj4653 commented:
@ncordon I've addressed your feedback and pushed the updates:

1. `Changelog Fix`: Resolved the validate-changelogs failure by escaping the brackets in the YAML summary.
2. `Test Coverage`: Restored and expanded tests for Greatest.java and Least.java in VerifierTests.java
- Verified support for all mapped types including String (Keyword/Text), IP, Version, and Date, in addition to Numerics.
- Specific tests for Text fields were added using fullTextAnalyzer.

Local build and tests passed. Ready for another look when you have a chance. Thanks!

Commit:
[CI] Auto commit changes from spotless

ncordon committed:
format

ncordon commented:
buildkite test this please

ncordon committed:
Merge branch 'main' into bugfix/114036-greatest-least-type-resolution

ncordon reviewed (approved):
Thanks for your contribution @shinsj4653. LGTM!
